tutorials/matchmaker101/matchfunction/mmf/matchfunction.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mmf
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"open-match.dev/open-match/pkg/matchfunction"
+	"open-match.dev/open-match/pkg/pb"
+)
+
+// This is a very basic Roster based Match Function. It expects the Roster(s) in
+// the input Match Profile to have the following format:
+//     Roster{
+//       Name: <Pool Name>,
+//       TicketIds: [EMPTY_ROSTER_SPOT, EMPTY_ROSTER_SPOT, ..]
+//     }
+// Where the Name maps to the Pool Name from which Tickets need to be populated
+// for a given Roster and TicketIds have number of 'EMPTY_ROSTER_SPOT' representing
+// the number of players to be populated for this Roster.
+//
+// The function then simply polls MMLogic for players for each Pool and populates
+// Roster slots with Tickets to matching the corresponding pool name. The generated
+// Match has all the Ticket IDs for the match and populated Roster(s).
+//
+// The Profiles generator in the tutorial generates Profiles that follow this format.
+const (
+	matchName       = "roster-based-matchfunction"
+	emptyRosterSpot = "EMPTY_ROSTER_SPOT"
+)
+
+// Run is this match function's implementation of the gRPC call defined in api/matchfunction.proto.
+func (s *MatchFunctionService) Run(req *pb.RunRequest, stream pb.MatchFunction_RunServer) error {
+	// Fetch tickets for the pools specified in the Match Profile.
+	poolTickets, err := matchfunction.QueryPools(stream.Context(), s.mmlogicClient, req.GetProfile().GetPools())
+	if err != nil {
+		log.Printf("Failed to query tickets for the given pools, got %w", err)
+		return err
+	}
+
+	// Generate proposals.
+	proposals, err := makeMatches(req.GetProfile(), poolTickets)
+	if err != nil {
+		log.Printf("Failed to generate matches, got %w", err)
+		return err
+	}
+
+	log.Printf("Streaming %v proposals to Open Match", len(proposals))
+	// Stream the generated proposals back to Open Match.
+	for _, proposal := range proposals {
+		if err := stream.Send(&pb.RunResponse{Proposal: proposal}); err != nil {
+			log.Printf("Failed to stream proposals to Open Match, got %w", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func makeMatches(p *pb.MatchProfile, poolTickets map[string][]*pb.Ticket) ([]*pb.Match, error) {
```

Comment:
This match logic seems complicated, the roster use doesn't seem intuitive to me.  Would be interested in discussing alternatives.

Reply:
Good feedback. I had ported it from a familiar case - but If it isnt intuitive to you, it aint gonna be intuitive for a 101 tutorial user. I've updated this to just pick a certain number from each pool for starters.

Reply:
This seems fine for a first iteration.  Thanks!